### ISSUE TITLE
Busy main thread causes bidders to be incorrectly timed out

### ISSUE THREAD

**Issue by snapwich** [prebid/Prebid.js#1046]
## Type of issue
Bug

## Description
The main problem here seems to be using Javascript timestamps to determine whether a bidder timed out, specifically in this line in `bidmanager#addBidResponse`

```
    if (bid.timeToRespond > $$PREBID_GLOBAL$$.cbTimeout + $$PREBID_GLOBAL$$.timeoutBuffer) {
      const timedOut = true;

      exports.executeCallback(timedOut);
    }
```
 
If the main thread is busy when responses are coming in (which it usually is) it is possible for a bidder that responded within the given auction time allotment to be marked as timed out.  We've found this to happen semi-frequently on publisher's pages who use very tight timeouts of 500ms or less.

## Steps to reproduce
An example test page to replicate this issue can be seen here: http://jsbin.com/gifusucika/1/edit?html,js,output

For the example I've emulated ajax responses with `setTimeout` for consistency, but confirmed that ajax responses behave similarly (having the response handler put on the event loop in order received, but not executed until main thread clears up).

I've emulated a busy main thread with a busy wait function called `sleep` and having the thread sleep before the end of the auction.

### Expected results
Since the slow response is received before the end of the auction time allotment it is expected that it would be included in the auction, especially since the auction timeout function has not fired yet (it's still in the event loop queue awaiting execution).

### Actual results
Due to the busy main thread, the `addBidResponse` fires later making its Javascript timestamp longer than the `bidderTimeout` + `timeoutBuffer`.  However, as stated before, the `addBidResponse` still fires before the auction timeout which gives the bid a chance to be included; but, since there is code checking the timestamps in javascript, the bid is excluded anyways.

<img width="690" alt="screen shot 2017-03-13 at 10 52 41 am" src="https://cloud.githubusercontent.com/assets/551085/23867988/66ac311a-07e3-11e7-8938-4c10e544ae5e.png">

## Platform details
Tested in Chrome, but should be the same across browsers.

## Other information
There is the 200ms buffer added to allow some leeway.  However, I would think a method that tests whether the auction is still live would be a more correct solution for determining whether a bid should be included over checking and comparing timestamps.  By virtue of the event loop's queue ordering we can know a bid came in before the end of the auction even if it was heavily delayed. 

Let me know your thoughts or if you think there is a quick fix.  I haven't attempted a fix yet because I'm not sure the intent of the "timed out" code and don't want to break things.

**Comment by mkendall07** [comment]
It's really just a way for us to end the auction if we are past the time allowed in the timeout. The reason the code exists was because of issues like this (https://github.com/prebid/Prebid.js/issues/311)  where bids are coming back with large timestamps based on when the JS executes.  Certainly would like to include bids that are already in the JS loop queue. I'm not sure how we could implement it better though.

**Comment by snapwich** [comment]
> It's really just a way for us to end the auction if we are past the time allowed in the timeout.

Isn't that what this is for: https://github.com/prebid/Prebid.js/blob/master/src/prebid.js#L417 or are there times when that is not enough the timeout case? 

If it is enough, then I think checking whether or not https://github.com/prebid/Prebid.js/blob/master/src/bidmanager.js#L103 actually returned something should be sufficient to determine if there is an active auction going on and if the current bid response is allowed to participate.

**Comment by mkendall07** [comment]
`setTimout` is very imprecise. So this is a way to control for it better. Open to suggestions though.

**Comment by protonate** [comment]
This is interesting, sourced from [here](https://www.sitepoint.com/creating-accurate-timers-in-javascript/).

```
function doTimer(length, resolution, oninstance, oncomplete)
{
    var steps = (length / 100) * (resolution / 10),
        speed = length / steps,
        count = 0,
        start = new Date().getTime();

    function instance()
    {
        if(count++ == steps)
        {
            oncomplete(steps, count);
        }
        else
        {
            oninstance(steps, count);

            var diff = (new Date().getTime() - start) - (count * speed);
            window.setTimeout(instance, (speed - diff));
        }
    }

    window.setTimeout(instance, speed);
}
```

We should also try to limit timers as each itself contends for the event loop.

**Comment by snapwich** [comment]
`setTimeout` _is_ imprecise, but most of that is due to either the busy main thread or intentional imprecisions when a tab is backgrounded.  The former will affect AJAX and JSONP requests all the same, the later I'm not sure; perhaps some testing is in order.

If AJAX response handlers are indeed more close to honoring "real-time", having them close the auction is probably not a bad idea.  However, they should probably do some form of "last call" perhaps by using something similar to `setImmediate` to allow all other AJAX handlers that have already responded to also register their bids before the auction is closed.  If AJAX handlers are similar to `setTimeout` in real-time honoring, then I suggest we just allow the `setTimeout` to close the auction and remove the ability of `addBidResponse` to close the auction.

@protonate the code you're listing is an incremental `setTimeout` used to smooth animations and such.  I don't think it's very relevant to this issue; or for animations, considering the availability of `requestAnimationFrame` now.

**Comment by mkendall07** [comment]
> However, they should probably do some form of "last call" perhaps by using something similar to setImmediate to allow all other AJAX handlers that have already responded to also register their bids before the auction is closed.

I agree, this is what we should do. It looks like `setTimeout(fn, 0);` approximates the same behavior as `setImmediate` since it's non-standard, along with `postMessage` so there should be a way to do this. Sounds like a fun PR :)

**Comment by protonate** [comment]
Also agree that `addBidResponse` shouldn't be responsible for closing the auction. Promises could be used here and is something we've considered for Prebid 1.0.

@snapwich the timeout handler above I referenced for using Date().getTime difference but of course we are already doing that. And you're right that the smoothing isn't relevant here.

**Comment by mkendall07** [comment]
oh, that sounds `promising` ;)

**Comment by hrgui** [comment]
> However, they should probably do some form of "last call" perhaps by using something similar to setImmediate to allow all other AJAX handlers that have already responded to also register their bids before the auction is closed.

What if the main thread is still busy during that time? Wouldn't there be a delay from the timeout to the last call?

In this example, https://gist.github.com/hrgui/a0397194a40b7962c2895b30514b388f, I executed a busy loop function after 4 ms, and I calculated the time difference between request start and the function execution which causes the request to be ended, When comparing the times from network to the JavaScript time, there is a significant delay (about 2 seconds). Without the busy loop, there is almost no delay (3ms) difference.
